### PR TITLE
Fix select by polygon highlight

### PR DESCRIPTION
**Pull request by vcloarec**
In the select by polygon tool, when layer's CRS and map's CRS are different, the highlight of selected polygon is wrong.
It comes from that the highlight needs the geometry in layer coordinates.
So need some coordinate transforms to make all good.

Before:
selectPolygonBefore

After:
selectPolygonAfter

### CODE REVIEW

File: src/app/qgsmaptoolselectionhandler.cpp [hunk 1]
```diff
@@ -263,9 +263,20 @@ void QgsMapToolSelectionHandler::selectPolygonPressEvent( QgsMapMouseEvent *e )
...
+          QgsCoordinateTransform transform = mCanvas->mapSettings().layerTransform( vectorLayer );
+
+          try
+          {
+            rect = transform.transformBoundingBox( rect, QgsCoordinateTransform::ReverseTransform );
+          }
+          catch ( QgsCsException &exception )
+          {
+            QgsDebugMsg( QStringLiteral( "Could not transform geometry to layer CRS" ) );
+          }
```
Comment (nyalldawson) [inline src/app/qgsmaptoolselectionhandler.cpp:276]: This shouldn't be required -- when setDestinationCrs is called on a feature request then the filter rect transformation is handled automatically

File: src/app/qgsmaptoolselectionhandler.cpp [hunk 2]
```diff
@@ -279,7 +290,20 @@ void QgsMapToolSelectionHandler::selectPolygonPressEvent( QgsMapMouseEvent *e )
...
     if ( !selectedFeatures.empty() && selectedFeatures[0].mFeature.hasGeometry() )
-      setSelectedGeometry( selectedFeatures[0].mFeature.geometry(), e->modifiers() );
+    {
+      QgsCoordinateTransform transform = mCanvas->mapSettings().layerTransform( selectedFeatures.at( 0 ).mLayer );
+      QgsGeometry geom = selectedFeatures[0].mFeature.geometry();
+      try
+      {
+        geom.transform( transform );
+      }
+      catch ( QgsCsException &exception )
```
Comment (nyalldawson) [inline src/app/qgsmaptoolselectionhandler.cpp:300]:
```suggestion
      catch ( QgsCsException & )
```

File: src/app/qgsmaptoolselectionhandler.cpp (hunk 1 above)
Comment (vcloarec) [reply src/app/qgsmaptoolselectionhandler.cpp:276]: yes, but here I want the result feature request in layer coordinate, so I don not use setDestinationCrs.

Comment (vcloarec) [reply src/app/qgsmaptoolselectionhandler.cpp]: @nyalldawson is there still a problem here?

Comment (nyalldawson) [reply src/app/qgsmaptoolselectionhandler.cpp]: I must admit I still don't follow the new logic -- I can't see anywhere in QgsIdentifyMenu where the crs is important, so I can't see why it's a problem to do the transformation once upfront.

I'm not doubting that this fixes the issue... but if I can't understand the code then it's a good sign that either API needs to be reworked or that we're missing a whole bunch of explanatory comments here (preferable the first!)

**Comment by vcloarec** [comment]
not stale

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
Thanks @vcloarec  -- I've looked into the code and agree with your logic!

**Pull request merged**